src/psyclone/psyir/transformations/inline_trans.py

```diff
@@ -0,0 +1,406 @@
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: A. R. Porter, STFC Daresbury Lab
+
+'''
+This module contains the InlineTrans transformation.
+
+'''
+
+from psyclone.errors import InternalError
+from psyclone.psyGen import Transformation
+from psyclone.psyir.nodes import (
+    Call, Routine, Reference, CodeBlock, Return, Literal)
+from psyclone.psyir.symbols import (ContainerSymbol, DataSymbol, ScalarType,
+                                    ImportInterface)
+from psyclone.psyir.transformations.transformation_error import (
+    TransformationError)
+
+
+class InlineTrans(Transformation):
+    '''
+    This transformation takes a Call and replaces it with the body of the
+    target routine. It is used as follows:
+
+    >>> from psyclone.psyir.frontend.fortran import FortranReader
+    >>> from psyclone.psyir.nodes import Call, Routine
+    >>> from psyclone.psyir.transformations import InlineTrans
+    >>> code = ("""
+    ... module test_mod
+    ... contains
+    ...   subroutine run_it()
+    ...   integer :: i
+    ...   real :: a(10)
+    ...   do i=1,10
+    ...     a(i) = 1.0
+    ...     call sub(a(i))
+    ...   end do
+    ...   end subroutine run_it
+    ...   subroutine sub(x)
+    ...     real, intent(inout) :: x
+    ...     x = 2.0*x
+    ...   end subroutine sub
+    ... end module test_mod"""
+    >>> psyir = FortranReader().psyir_from_source(code)
+    >>> call = psyir.walk(Call)[0]
+    >>> inline_trans = InlineTrans()
+    >>> inline_trans.apply(call)
+    >>> # Uncomment the following line to see a text view of the schedule
+    >>> # print(psyir.walk(Routine)[0].view())
+
+    The routine to be inlined must not contain any un-resolved symbols or
+    access any symbols that are defined in the parent container. Currently,
+    (#924) the routine must also be in the same source file as the call.
+
+    '''
+    def apply(self, node, options=None):
+        '''
+        Takes the body of the routine that is the target of the supplied
+        call and replaces the call with it.
+
+        :param node: target PSyIR node.
+        :type node: subclass of :py:class:`psyclone.psyir.nodes.Routine`
+        :param options: a dictionary with options for transformations.
+        :type options: Optional[Dict[str, str]]
+
+        '''
+        self.validate(node, options)
+
+        # The table we will copy symbols into.
+        table = node.scope.symbol_table
+        # Find the routine to be inlined.
+        orig_routine = self._find_routine(node)
+
+        if not orig_routine.children or isinstance(orig_routine.children[0],
+                                                   Return):
+            # Called routine is empty so just remove the call.
```

Comment:
What would happen if this would be a function that only returned a value? I haven't seen any functions in the tests, have these been considered?

Reply:
Ah, not until now!

Reply:
To keep things simple for now, I've just extended validate to reject this case.